packages/by-name/kata/kata-runtime/0020-genpolicy-support-dynamic-annotations.patch

```diff
@@ -0,0 +1,177 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Moritz Sanft <58110325+msanft@users.noreply.github.com>
+Date: Fri, 17 Jan 2025 16:39:29 +0100
+Subject: [PATCH] genpolicy: support dynamic annotations
+
+This adds support for handling annotations with dynamic keys to
+genpolicy. This is necessary for use-cases like GPU containers, where
+in-cluster components (i.e. post policy-generation) instrument
+containers with annotations with varying keys, like `cdi.k8s.io/vfioXY`,
+where `XY` corresponds to a dynamic ID.
+---
+ src/tools/genpolicy/genpolicy-settings.json |  8 ++-
+ src/tools/genpolicy/rules.rego              | 55 +++++++++++++++++++--
+ src/tools/genpolicy/src/policy.rs           | 10 ++++
+ 3 files changed, 67 insertions(+), 6 deletions(-)
+
+diff --git a/src/tools/genpolicy/genpolicy-settings.json b/src/tools/genpolicy/genpolicy-settings.json
+index 9b95f9f7462717d04f0b9ce685d97c0455f949da..50406e8ac7a961d317f2254f30339c0c1d399f18 100644
+--- a/src/tools/genpolicy/genpolicy-settings.json
++++ b/src/tools/genpolicy/genpolicy-settings.json
+@@ -137,6 +137,9 @@
+             "io.katacontainers.pkg.oci.container_type": "pod_container",
+             "io.kubernetes.cri.container-type": "container"
+         },
++        "DynamicAnnotations": {
++            "^cdi\\.k8s\\.io\\/vfio[0-9]{2}$": "^nvidia.com/gpu=[0-9]+$"
++        },
+         "Linux": {
+             "Sysctl": {
+                 "net.ipv4.ip_unprivileged_port_start": "0",
+@@ -333,7 +336,8 @@
+                 "^AZURE_CLIENT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_TENANT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_FEDERATED_TOKEN_FILE=/var/run/secrets/azure/tokens/azure-identity-token$",
+-                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$"
++                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$",
++                "^PCI_RESOURCE_NVIDIA_COM.*=[a-fA-F0-9:.-]*$"
+             ]
+         },
+         "CopyFileRequest": [
+@@ -348,4 +352,4 @@
+         "UpdateEphemeralMountsRequest": false,
+         "WriteStreamRequest": false
+     }
+-}
+\ No newline at end of file
++}
+diff --git a/src/tools/genpolicy/rules.rego b/src/tools/genpolicy/rules.rego
+index 43cb19a56fe8ea5833708f0639c9e85ddd884cb3..554ab211b1392f5163cb894cfc29bc17cb4012d0 100644
+--- a/src/tools/genpolicy/rules.rego
++++ b/src/tools/genpolicy/rules.rego
+@@ -152,10 +152,10 @@ state_allows(key, value) = action {
+   state := get_state()
+   not state[key]
+   print("state_allows: saving to state key =", key, "value =", value)
+-  path := get_state_path(key) 
++  path := get_state_path(key)
+   action := {
+     "op": "add",
+-    "path": path, 
++    "path": path,
+     "value": value,
+   }
+ }
+@@ -199,13 +199,17 @@ allow_anno(p_oci, i_oci) {
+ }
+ allow_anno(p_oci, i_oci) {
+     print("allow_anno 2: p Annotations =", p_oci.Annotations)
++    print("allow_anno 2: p DynamicAnnotations =", p_oci.DynamicAnnotations)
```

Comment:
Does this already fail if there is no `DynamicAnnotations` on `p_oci`?

Reply:
No, this will be `undefined` by default.